### ISSUE TITLE
No error when subclass does not provide a value for ClassVar

### ISSUE THREAD

**Issue by saulbein** [python/mypy#14381]
**Bug Report**

I've been trying to mypy to typecheck subclasses that need to provide a required value for a `ClassVar` in a base class that cannot provide the value itself (only has an unintialized attribute marker). The problem is that badly implemented subclasses can ignore this attribute without mypy raising errors about it (though interestingly if the value provided has incorrect type, mypy will raise errors about _that_).

**To Reproduce**
```python
from abc import ABC
from typing import ClassVar, Type


class Base(ABC):
    my_super_important_value: ClassVar[str]


class GoodSubclass(Base):
    my_super_important_value = "1337"


# I want mypy to error about an unimplemented variable but it doesn't
class BadSubclass(Base):
    pass


def print_abc(cls: Type[Base]) -> None:
    print(cls.my_super_important_value)


print_abc(GoodSubclass)
# 1337
print_abc(BadSubclass)
# AttributeError: type object 'BadSubclass' has no attribute 'my_super_important_value'
```
[mypy playground](https://mypy-play.net/?mypy=latest&python=3.11&flags=strict%2Ccheck-untyped-defs&gist=0f863f2155861ed2ea35a2fd1194a3a5)

**Expected Behavior**

Expected `mypy` to show an error about the `BadSubclass` class not providing a value for `my_super_important_value`. 

**Actual Behavior**

No errors.

**Your Environment**

- Mypy version used: 0.991
- Mypy command-line flags: --strict
- Mypy configuration options from `mypy.ini` (and other config files): N/A
- Python version used: 3.10

Note that the same code fails with an error when [using the Pyre type checker](https://pyre-check.org/play?input=from%20abc%20import%20ABC%0Afrom%20typing%20import%20ClassVar%2C%20Type%0A%0A%0Aclass%20Base(ABC)%3A%0A%20%20%20%20my_super_important_value%3A%20ClassVar%5Bstr%5D%0A%0A%0Aclass%20GoodSubclass(Base)%3A%0A%20%20%20%20my_super_important_value%20%3D%20%221337%22%0A%0A%0A%23%20I%20want%20mypy%20to%20error%20about%20an%20unimplemented%20variable%20but%20it%20doesn%27t%0Aclass%20BadSubclass(Base)%3A%0A%20%20%20%20pass%0A%0A%0Adef%20print_abc(cls%3A%20Type%5BBase%5D)%20-%3E%20None%3A%0A%20%20%20%20print(cls.my_super_important_value)%0A%0A%0Aprint_abc(GoodSubclass)%0A%23%201337%0Aprint_abc(BadSubclass)%0A%23%20AttributeError%3A%20type%20object%20%27BadSubclass%27%20has%20no%20attribute%20%27my_super_important_value%27%0A) with this error:
```
14:0: Uninitialized attribute [13]: Attribute `my_super_important_value` inherited from abstract class `Base` in class `BadSubclass` to have type `str` but is never initialized. 
```

**Comment by thomkeh** [comment]
I have a PR to fix this, #13797, but it doesn't look like it'll get merged.

**Comment by bwo** [comment]
Bump on this! It would be very useful.

Note that you can *almost* accomplish something like this this way:

```python
from typing import ClassVar
from abc import abstractmethod


class Foo:

    @classmethod
    @property
    @abstractmethod
    def x(cls) -> str:
        pass


    @property
    @abstractmethod
    def y(self) -> str:
        pass

class Bar(Foo):
    y = 'hi' 


print(Bar().y)  # error: Cannot instantiate abstract class "Bar" with abstract attribute "x"  [abstract]
```

But mypy will *also* complain that "Only instance methods can be decorated with @property", and it will also accept an instance-level property definition for `x`.

@sobolevn is anything blocking @tmke8's PR?

**Comment by asottile** [comment]
here's ~slightly closer to a solution -- mypy will prevent instantiation:

```python
from typing import Protocol, ClassVar

class _Base(Protocol):
    name: ClassVar[str]

class Abstract(_Base): ...

class ConcreteGood(Abstract):
    name = 'good!'

class ConcreteBad(Abstract): ...  # bad!  missing name!

print(ConcreteGood().name)  # ok
print(ConcreteBad().name)  # mypy will raise an error here!
```

but it doesn't prevent usage of `ConcreteBad` as a value for `type[Abstract]` which would be nice
